cmd/lambda/accounts/update.go

```diff
@@ -2,116 +2,53 @@ package main
-// (ie. they don't want to update it)
-// or, they explicitly set it to an empty value
-type updateAccountRequest struct {
-	ID                  *string                 `json:"id"`
-	AccountStatus       *db.AccountStatus       `json:"accountStatus"`
-	LastModifiedOn      *int64                  `json:"lastModifiedOn"`
-	CreatedOn           *int64                  `json:"createdOn"`
-	AdminRoleArn        *string                 `json:"adminRoleArn"`        // Assumed by the master account, to manage this user account
-	PrincipalRoleArn    *string                 `json:"principalRoleArn"`    // Assumed by principal users
-	PrincipalPolicyHash *string                 `json:"principalPolicyHash"` // The policy used by the PrincipalRoleArn
-	Metadata            *map[string]interface{} `json:"metadata"`
-}
-
+// UpdateAccountByID updates an accounts information based on ID
 func UpdateAccountByID(w http.ResponseWriter, r *http.Request) {
 	accountID := mux.Vars(r)["accountId"]
 
 	// Deserialize the request JSON as an request object
-	var request *updateAccountRequest
+	d := model.Account{}
 	decoder := json.NewDecoder(r.Body)
-	err := decoder.Decode(&request)
+	err := decoder.Decode(&d)
 	if err != nil {
-		response.WriteAPIErrorResponse(w, http.StatusBadRequest, "ClientError", "invalid request parameters")
+		api.WriteAPIErrorResponse(w,
+			errors.NewBadRequest("invalid request parameters"))
 		return
 	}
-	request.ID = &accountID
 
-	// If the request includes a new adminRoleArn,
-	// validate that we can assume the ARN
-	if request.AdminRoleArn != nil {
-		_, err = TokenSvc.AssumeRole(&sts.AssumeRoleInput{
-			RoleArn:         request.AdminRoleArn,
-			RoleSessionName: aws.String("MasterAssumeRoleVerification"),
-		})
-
-		if err != nil {
-			response.WriteRequestValidationError(
-				w,
-				fmt.Sprintf("Unable to update account %s: "+
-					"admin role is not assumable by the master account",
-					accountID),
-			)
-			return
-		}
+	var dataSvc dataiface.AccountData
+	if err := Services.Config.GetService(&dataSvc); err != nil {
+		api.WriteAPIErrorResponse(w, err)
+		return
 	}
 
-	// Figure out which DB fields we want to update
-	// Note that not all fields are updatable
-	// eg, we don't want clients to set their own timestamps,
-	// 	or modify fields, like AccountStatus, which are  used for
-	// 	internal bookkeeping.
-	var fieldsToUpdate []string
-	accountPartial := db.Account{
-		ID: *request.ID,
-	}
-	if request.AdminRoleArn != nil {
-		fieldsToUpdate = append(fieldsToUpdate, "AdminRoleArn")
-		accountPartial.AdminRoleArn = *request.AdminRoleArn
-	}
-	if request.Metadata != nil {
-		fieldsToUpdate = append(fieldsToUpdate, "Metadata")
-		accountPartial.Metadata = *request.Metadata
-	}
-	if len(fieldsToUpdate) == 0 {
-		response.WriteRequestValidationError(
-			w,
-			fmt.Sprintf("Unable to update account %s: "+
-				"no updatable fields provided",
-				accountID),
-		)
+	var amSvc accountmanageriface.AccountManagerAPI
```

Comment:
So back were looking at the account data layer, there was some discussion around the pattern of injecting service dependencies at the method level, versus injecting them on the service structs.

The thing about having to pass in dependencies at the method level is that there's a lot of low-level service configuration and wiring that now needs to happen in the controller itself -- which is how we end up with all these "Account" services in the code, even though I'm really only using one of them.

I'd like if we could get to the point where our controller looked a lot thinner, something just like:

```golang
func UpdateAccountByID(w http.ResponseWriter, r *http.Request) {
	accountID := mux.Vars(r)["accountId"]

	// Deserialize the request JSON as an request object
	accountUpdates := model.Account{}
	decoder := json.NewDecoder(r.Body)
	err := decoder.Decode(&accountUpdates)
	if err != nil {
		api.WriteAPIErrorResponse(w,
			errors.NewBadRequest("invalid request parameters"))
		return
	}

        // Update the account
	updatedAccount, err := config.Accounts.Update(accountUpdates)
	if err != nil {
		api.WriteAPIErrorResponse(w, err)
		return
	}

	api.WriteAPIResponse(w, http.StatusOK, updatedAccount)
}

Reply:
Yep been trying to stick with the making interfaces smaller mantra.  I actually don't mind the idea of passing in interfaces to functions.   However, since we have greatly simplified the data layer what we have can be considered smaller.  So we can adjust if you would prefer.